### PR TITLE
[KYUUBI #988] Add Kyuubi session event

### PR DESCRIPTION
**Pull request by ulysses-you**
### _Why are the changes needed?_

closes https://github.com/apache/incubator-kyuubi/issues/988

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -735,4 +735,13 @@ object KyuubiConf {
       .version("1.3.0")
       .booleanConf
       .createWithDefault(true)
+
+  val SESSION_NAME: OptionalConfigEntry[String] =
+    buildConf("session.name")
+      .doc("Session name is used to relate the history session which can help optimize. " +
+        "Please set this config using different value for workflows which are session level. " +
+        "This config is none means we can not find the history session.")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:743]: Session name is just a session name. History is just a typical use case, the doc is overkilling to me

Comment (ulysses-you) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: how about this updating

File: kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  def apply(session: KyuubiSessionImpl): KyuubiSessionEvent = {
+    assert(KyuubiServer.kyuubiServer.isDefined)
+    val serverIP = KyuubiServer.kyuubiServer.get.connectionUrl
+    KyuubiSessionEvent(
+      session.handle.identifier.toString,
+      session.sessionName,
+      session.user,
+      session.ipAddress,
+      serverIP,
+      session.sessionConf.getAll,
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:59]: How about recording the `session.conf` only

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 3]
```diff
@@ -50,6 +53,10 @@ class KyuubiSessionImpl(
 
   private val engine: EngineRef = EngineRef(sessionConf, user, handle)
 
+  val sessionName: String = sessionConf.getOption(KyuubiConf.SESSION_NAME.key).getOrElse("")
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:56]: how about move this inside the KyuubiSessionEvent.apply?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 4]
```diff
@@ -50,6 +53,10 @@ class KyuubiSessionImpl(
...
+  private val sessionEvent = KyuubiSessionEvent.apply(this)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:57]: how about KyuubiSessionEvent(this)

File: kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    val sessionName: String =
+      session.sessionConf.getOption(KyuubiConf.SESSION_NAME.key).getOrElse("")
+    KyuubiSessionEvent(
+      session.handle.identifier.toString,
+      sessionName,
+      session.user,
+      session.ipAddress,
+      serverIP,
+      session.handle.protocol.getValue,
+      session.normalizedConf,
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:66]: how about track the raw conf?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  override def partitions: Seq[(String, String)] =
+    ("day", Utils.getDateFromTimestamp(startTime)) :: Nil
+}
+
+object KyuubiSessionEvent {
+  def apply(session: KyuubiSessionImpl): KyuubiSessionEvent = {
+    assert(KyuubiServer.kyuubiServer.isDefined)
+    val serverIP = KyuubiServer.kyuubiServer.get.connectionUrl
+    val sessionName: String =
+      session.sessionConf.getOption(KyuubiConf.SESSION_NAME.key).getOrElse("")
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala:58]: how about get it from session.conf

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiServer.scala [hunk 7]
```diff
@@ -32,6 +32,7 @@ import org.apache.kyuubi.zookeeper.EmbeddedZookeeper
 
 object KyuubiServer extends Logging {
   private val zkServer = new EmbeddedZookeeper()
+  private[kyuubi] var kyuubiServer: Option[KyuubiServer] = None
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiServer.scala:35]: private[kyuubi] var kyuubiServer: KyuubiServer = _

File: kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala (hunk 6 above)
Comment (ulysses-you) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/events/KyuubiSessionEvent.scala]: yeah, also update the config docs to clarify `we only apply this value from session conf`.

**Review by yaooqinn — APPROVED** [review]

**Comment by ulysses-you** [comment]
thank you for review, merging to master for v1.4.0

**Pull request closed without merging**